[SPARK-53535][SQL] Fix missing structs always being assumed as nulls

ZiyaZa opened the pull request:
### What changes were proposed in this pull request?

Currently, if all fields of a struct mentioned in the read schema are missing in a Parquet file, the reader populates the struct with nulls.

This PR modifies the scan behavior so that if the struct exists in the Parquet schema but none of the fields from the read schema are present, we instead pick an arbitrary field from the Parquet file to read and use that to populate NULLs (as well as outer NULLs and array sizes if the struct is nested in another nested type).

This is done by changing the schema requested by the readers. We add an additional field to the requested schema when clipping the Parquet file schema according to the Spark schema. This means that the readers actually read and return more data than requested, which can cause problems. This is only a problem for the `VectorizedParquetRecordReader`, since for the other read code path via parquet-mr, we already have an `UnsafeProjection` for outputting only requested schema fields in `ParquetFileFormat`.

To ensure `VectorizedParquetRecordReader` only returns Spark requested fields, we create the `ColumnarBatch` with vectors that match the requested schema (we get rid of the additional fields by recursively matching `sparkSchema` with `sparkRequestedSchema` and ensuring structs have the same length in both). Then `ParquetColumnVector`s are responsible for allocating dummy vectors to hold the data temporarily while reading, but these are not exposed to the outside.

The heuristic to pick the arbitrary leaf field is as follows: We try to minimize the amount of arrays or maps (repeated fields) in the path to a leaf column, because the more repeated fields we have the more likely we are to read larger amount of data. At the same repetition level, we consider the type of each column to pick the cheapest column to read (struct nesting do not affect the decision here). We look at the byte size of the column type to pick the cheapest one as follows:
    - BOOLEAN: 1 byte
    - INT32, FLOAT: 4 bytes
    - INT64, DOUBLE: 8 bytes
    - INT96: 12 bytes
    - BINARY, FIXED_LEN_BYTE_ARRAY, default case for future types: 32 bytes (high cost due to variable/large size)

### Why are the changes needed?

This is a bug fix, because we were incorrectly assuming non-null struct values to be missing from the file depending on requested fields and returning null values.

### Does this PR introduce _any_ user-facing change?

Yes. We previously assumed structs to be null if all the fields we are trying to read from a Parquet file were missing from that file, even if the file contained other fields that could be used to take definition levels from. See an example from the Jira ticket below:

```python
df_a = sql('SELECT 1 as id, named_struct("a", 1) AS s')
path = "/tmp/missing_col_test"
df_a.write.format("parquet").save(path)

df_b = sql('SELECT 2 as id, named_struct("b", 3) AS s')
spark.read.format("parquet").schema(df_b.schema).load(path).show()
```

This used to return:

```
+---+----+
| id|   s|
+---+----+
|  1|NULL|
+---+----+
```

It now returns:

```
+---+------+
| id|     s|
+---+------+
|  1|{NULL}|
+---+------+
```

### How was this patch tested?

Added new unit tests, also fixed an old test to expect this new behavior.

### Was this patch authored or co-authored using generative AI tooling?

No.

ZiyaZa committed:
Fix missing structs always being assumed as nulls

ZiyaZa committed:
Fix linter errors

ZiyaZa committed:
Ensure sparkRequestedSchema is filled

juliuszsompolski reviewed (approved):
@gengliangwang could you take a look?

ZiyaZa committed:
Add a flag

ZiyaZa committed:
Merge branch 'master' into missing_struct

ZiyaZa commented:
@gengliangwang 
> I am not a big fan of such behavior changes. 

This is a behavior change that is required to fix a correctness issue. The issue is described in more detail in the linked [JIRA ticket](https://issues.apache.org/jira/browse/SPARK-53535). According to the comment https://github.com/apache/spark/pull/52557#issuecomment-3387926488 above, we could also get NullPointerException if the struct is marked as nullable, because we would wrongly assume all struct values to be null previously.

> why just picking one arbitrary field, instead of setting all the fields null?

We need to understand **for each row** if the struct value is null or it is a struct with all the fields as null (to explain in JSON notation, `null` and `{ 'a': null }` are not the same thing). The only way we can understand this is by looking at a child field of a struct that is present in the file, because Parquet stores nullability information in the definition levels of leaf columns. Based on that definition levels, we can identify in which rows struct is null or non-null.

> Could you provide more details on this one?

Updated the description.

> For breaking changes, usually we should introduce a SQL configuration to control the new/legacy behaviors

Added a flag to control this behavior.

> also we should update https://spark.apache.org/docs/latest/sql-migration-guide.html

Can you please explain how we update that? It looks like that is built from the `sql-migration-guide.md` file, but it currently does not contain anything for the next release if I read it correctly.

gengliangwang commented:
> It looks like that is built from the sql-migration-guide.md file, but it currently does not contain anything for the next release if I read it correctly.

Yes, please create a new section `## Upgrading from Spark SQL 4.0 to 4.1`

ZiyaZa committed:
Change to a legacy flag

ZiyaZa committed:
Add migration guide

ZiyaZa committed:
Split a long line

ZiyaZa commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1534,6 +1534,16 @@ object SQLConf {
       .createWithDefault(true)
 
+  val PARQUET_READ_ANY_FIELD_FOR_MISSING_STRUCT =
+    buildConf("spark.sql.parquet.readAnyFieldForMissingStruct")
```
I named it `spark.sql.legacy.parquet.returnNullStructIfAllFieldsMissing` to show that it is for Parquet files.

cloud-fan commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java:
```diff
@@ -265,7 +265,13 @@ private void initBatch(
+        .defaultValue().get());
+    StructType batchSchema = !returnNullStructIfAllFieldsMissing
+      ? new StructType(sparkSchema.fields())
+      : (StructType) truncateType(sparkSchema, sparkRequestedSchema);
```
the code before is `StructType batchSchema = new StructType(sparkSchema.fields());`, but now we only go to it if legacy config is off?

cloud-fan replied:
ah I see, we need to call `truncateType` for the legacy code path, to deal with the extra cheap field we picked when all fields are missing. Can we add code comments to explain it?

cloud-fan commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java:
```diff
@@ -309,6 +316,58 @@ public void initBatch(StructType partitionColumns, InternalRow partitionValues)
+   * requestedType: array<struct<a:int,b:long>>
+   * returns:       array<struct<a:int,b:long>>
+   * }</pre>
+   * We cannot return requestedType here because there might be slight differences, like nullability
```
shall we choose a different example? the current example simplify returns `requestedType`

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetReadSupport.scala:
```diff
@@ -418,11 +428,16 @@ object ParquetReadSupport extends Logging {
         parquetRecord.getFields.asScala.groupBy(_.getName.toLowerCase(Locale.ROOT))
     lazy val idToParquetFieldMap =
         parquetRecord.getFields.asScala.filter(_.getId != null).groupBy(f => f.getId.intValue())
+    var isStructWithMissingFields = true
```
```suggestion
    var isStructWithMissingAllFields = true
```

ZiyaZa committed:
Address review comments

cloud-fan commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java:
```diff
@@ -265,7 +265,15 @@ private void initBatch(
+    StructType batchSchema = returnNullStructIfAllFieldsMissing
+      ? new StructType(sparkSchema.fields())
+      // Truncate to match requested schema to make sure extra struct field that we read for
+      // nullability is not included in columnarBatch and exposed outside.
```
@ZiyaZa do we have test coverage for it? The tests passed both before and after you flip `returnNullStructIfAllFieldsMissing` here...

ZiyaZa committed:
Test for scan output not containing additional columns

ZiyaZa committed:
Empty

ZiyaZa commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala:
```diff
@@ -755,10 +763,72 @@ class ParquetIOSuite extends QueryTest with ParquetTest with SharedSparkSession
-          )
+          val df = spark.read.schema(readSchema).parquet(file)
+          val scanNode = df.queryExecution.executedPlan.collectLeaves().head
+          VerifyNoAdditionalScanOutputExec(scanNode).execute().first()
```
I tried to use that first, but couldn't get it to work because `ColumnarToRowExec` adds its own `UnsafeProjection` with `child.output`. Let's keep the custom plan node.

ZiyaZa committed:
Simplify custom test plan node

ZiyaZa committed:
Empty

cloud-fan commented:
thanks, merging to master!
